includes/class-monk-links.php

```diff
@@ -0,0 +1,523 @@
+	}
+
+	/**
+	 * Checks whether the pretty permalinks are active or not
+	 * if it is, return the structure
+	 *
+	 * @return bool|string $structure
+	 */
+	public function monk_using_permalinks() {
+		$structure = get_option( 'permalink_structure', false );
+		return ( empty( $structure ) ) ? false : $structure;
+	}
+
+	/**
+	 * Function to check for the usage of a trailing slash
+	 *
+	 * @return bool
+	 */
+	public function monk_apply_slash() {
+		if ( $this->monk_using_permalinks() ) {
+			$structure = $this->monk_using_permalinks();
+			return ( '/' === substr( $structure, -1, 1 ) ) ? true : false;
+		}
+	}
+
+	/**
+	 * Changes the link to the home page using the current language.
+	 *
+	 * @since  0.2.0
+	 * @param  string $url home link during query.
+	 * @param  string $path path requested.
+	 * @return string $url.
+	 */
+	public function monk_add_language_home_permalink( $url, $path ) {
+
+		// these cases are exceptions, do not filter.
+		if ( is_admin() || ! ( did_action( 'login_init' ) || did_action( 'template_redirect' ) ) ) {
+			return $url;
+		}
+
+		$site_language = get_option( 'monk_default_language', false );
+		$url_language  = get_query_var( 'lang' );
+		$language      = ( empty( $url_language ) ) ? $site_language : $url_language;
+
+		if ( $language && '/' === $path ) {
+			if ( $this->monk_using_permalinks() ) {
+				return trailingslashit( trailingslashit( $url ) . $language );
+			} else {
+				return add_query_arg( 'lang', $language, $url );
+			}
+		}
+
+		return $url;
+	}
+
+	/**
+	 * This function performs a redirect from the home url
+	 * when the user try to access it with no language
+	 *
+	 * @since  0.2.0
+	 * @return void
+	 */
+	public function monk_redirect_home_url() {
+		// url is wether site.com/en or site.com/?lang=en.
+		if ( is_home() && ! ( get_query_var( 'lang' ) || get_query_var( 's' ) ) ) {
+
+			$site_language = get_option( 'monk_default_language', false );
+			$url_language  = get_query_var( 'lang' );
+			$language      = ( empty( $url_language ) ) ? $site_language : $url_language;
+
+			if ( $this->monk_using_permalinks() ) {
+				wp_safe_redirect( trailingslashit( home_url( '/' . $language ) ) );
+				exit();
+			} else {
+				wp_safe_redirect( add_query_arg( 'lang', $language, trailingslashit( home_url() ) ), 301 );
+				exit();
+			}
+		} else {
+			return;
+		}
+	}
+
+	/**
+	 * Changes the post permalinks to add its language
+	 *
+	 * @since  0.2.0
+	 * @param  string $permalink Post link during query.
+	 * @param  object $post Post object.
+	 * @return string $permalink
+	 */
+	public function monk_add_language_post_permalink( $permalink, $post ) {
+		global $wp_rewrite;
+
+		$site_language = get_option( 'monk_default_language', false );
+		$post_language = get_post_meta( $post->ID, '_monk_post_language', true );
+		$url_language  = get_query_var( 'lang' );
+		$language      = ( empty( $post_language ) ) ? $site_language : $post_language;
+
+		if ( empty( $language ) ) {
```

Comment:
You need return a permalink always or you can broke the functionality.
You must encapsulate this change language process in another method, you're using it a lot in the methods below.